build_runner_core/lib/src/logging/log_renderer.dart

```diff
@@ -0,0 +1,83 @@
+// Copyright (c) 2025, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'package:build/build.dart';
+import 'package:crypto/crypto.dart';
+
+/// Renders objects to strings for display in logs.
+class LogRenderer {
+  final String rootPackageName;
+
+  LogRenderer({required this.rootPackageName});
+
+  /// Renders [id].
+  ///
+  /// Like `AssetId.toString`, except the package name is omitted if it matches
+  /// [rootPackageName].
+  String id(AssetId id) {
+    if (id.package == rootPackageName) {
+      return id.path;
+    } else {
+      return id.toString();
+    }
+  }
+
+  /// Renders a build action with [input] and [outputs].
+  ///
+  /// Output names that mostly match [input] are shortened to remove the
+  /// duplication.
+  ///
+  /// Examples: `foo.dart->.g.dart`, `foo.dart->(.g.dart, .other.dart)`
+  String build(AssetId input, Iterable<AssetId> outputs) {
+    final displayedInput = id(input);
+    // If the input ends in `.dart`, take the base name and trim it from all
+    // outputs. So for example `foo.dart->foo.g.dart` is shortened to
+    // `foo.dart->.g.dart`.
+    final trimPrefix =
+        displayedInput.endsWith('.dart')
+            ? displayedInput.substring(0, displayedInput.length - 5)
+            : null;
+    final displayedOutputs =
+        outputs.length == 1
+            ? _trim(id(outputs.first), trimPrefix: trimPrefix)
+            : '(${trimmedIds(outputs, trimPrefix: trimPrefix)})';
+    return '$displayedInput->$displayedOutputs';
+  }
+
+  /// Returns [ids] converted to `String` as a comma-separated list,
+  /// with all ids beyond the third replaced by "(# more)".
+  ///
+  /// If [trimPrefix] is passed, trim the specified prefix from all items where
+  /// it matches.
+  String trimmedIds(Iterable<AssetId> ids, {String? trimPrefix}) {
+    final firstThree = ids
+        .map(id)
+        .take(3)
+        .map((s) => _trim(s, trimPrefix: trimPrefix));
+    if (firstThree.length == ids.length) {
+      return firstThree.join(', ');
+    } else {
+      return [...firstThree, '(${ids.length - 3} more)'].join(', ');
```

Comment:
Doesn't have to be in this PR (or at all) but I always dislike seeing stuff like "a, b, c and 1 more" --- it would have been shorter and more informative to just say "a, b, c, d".

Reply:
Me too :)

I _think_ this is a good tradeoff, happy to tweak in future if it turns out not to be :)

Build outputs: mostly a build action has one output per input, so it won't hit the limit. This is logged a _lot_ of times, often multiple times per file in the main package, so I like having the limit just in case someone manages to end up with lots of outputs.

Files that changed: usually it will be exactly one file that the user just added / modified / changed, which won't hit the limit. But there's also the case that they changed git branch or did a "pub get" where there may be a large list of files; I wanted to avoid logging the whole list for that case as it's probably not interesting.